### ISSUE TITLE
Getting `error: too many basic types` when trying to import efi.h

### ISSUE THREAD

**Issue by typoon** [radareorg/radare2#16686]
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Ubuntu x86_64
| File format of the file you reverse (mandatory)      | PE32+ (EFI)
| Architecture/bits of the file (mandatory)            | x86/64 (EFI)
| r2 -v full output, **not truncated** (mandatory)     | Cutter-v1.10.3-pre-x64.Linux.AppImage

### Expected behavior

When trying to load the `efi.h` file, I am getting the following error:

~~~
[0x000102a0]> e dir.types=/usr/include/efi/x86_64

[0x000102a0]> to /usr/include/efi/efi.h

#include "/usr/include/efi/x86_64/efibind.h"
In file included from /usr/include/efi/efi.h:41:
/usr/include/efi/x86_64/efibind.h:78: error: too many basic types
In file included from /usr/include/efi/efi.h:41:
/usr/include/efi/x86_64/efibind.h:78: error: declaration expected
~~~

### Actual behavior

I was expecting the types declared in the file to be properly loaded

### Steps to reproduce the behavior 

1. Extract the `promptservice.zip` file attached and load it into radare.
2. Install the gnu EFI headers (sudo apt install gnu-efi). If you are not using ubuntu, use the headers that I included as an attachment. Extract them to `/usr/include/efi`
3. Enter the following commands in radare:

```
e dir.types=/usr/include/efi/x86_64
to /usr/include/efi/efi.h
```

4. Notice the error about too many basic types and declaration expected.

[efi.zip](https://github.com/radareorg/radare2/files/4517335/efi.zip)
[promptservice.zip](https://github.com/radareorg/radare2/files/4517336/promptservice.zip)

**Comment by radare** [comment]
thats totally expected, you must create your custom .h files to load into r2, we dont use gcc/clang so there will be many missing defines and such.

**Comment by typoon** [comment]
In this case, I am reversing a UEFI binary. I need to have those type
definitions from the headers, and it would be really hard to build a file
with them. Also, if some definitions change between versions, the I would
have to recreate a new header because of that instead of using the existing
ones.

Isn't there a way to load the headers like that, and when r2 runs into a
type it already knows, it either ignores or replaces the type based on a
configuration var? I believe that would be super helpful in many cases.

**Comment by radare** [comment]
Yes. The parser improves over time. But a more concrete reproducer would be better. So having the reason why a specific construction fails is the best way to solve this. Because there may be more than one issue probably.

I didnt checked the files but i assume you can probably comment regions with #if 0 until you get it to load

**Comment by typoon** [comment]
The issue here seems to be for trying to re-implement a type that radare already knows (or is hard coded, not sure). For example, if you create a `test.h` file with the following contents:

```
typedef unsigned int xpto_t;
typedef unsigned int xpto_t2;
```

And then try to load it with `to`, it works fine. Now, change the file to look like this:

```
typedef unsigned int xpto_t;
typedef unsigned int xpto_t2;
typedef unsigned int uint32_t;
```

And you get an error on the last typedef there:

~~~
/home/gilgamesh/tmp/efi/test.h:3: error: identifier expected
/home/gilgamesh/tmp/efi/test.h:3: error: declaration expected
~~~

So the issue seems to be that radare doesn't like it when it sees a type that it already knows internally. An I say `internally` because this here works fine:

```
typedef unsigned int xpto_t;
typedef unsigned int xpto_t2;
typedef unsigned int xpto_t;
```

As you can see, the `xpto_t` is defined twice, but radare is ok with that. Issue seems to be only with the types it considers `primitive`. My understanding is that the `*_t` types aren't really primitive (at least, according to the C spec <http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1124.pdf> section 6.2.5 and 6.7.2). So maybe the parser should allow those types to be overwritten.

Hopefully this helped clarify the issue.

**Comment by XVilka** [comment]
> My understanding is that the *_t types aren't really primitive (at least, according to the C spec http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1124.pdf section 6.2.5 and 6.7.2). So maybe the parser should allow those types to be overwritten.

Not exactly. While they are primitive for C some of the really "primitive" types like `int` or `long` are platform-dependent. And radare2 on 32bit ARM FreeBSD should be able to define structs for x64 WIndows PE files, which has a different understanding of `int` length. Thus, the `intX_t` files are considered primitive and the interpretation of the `int` by `to` depends on the current value of `e asm.bits` and `e asm.arch`, etc. So I think redifing `uint32_t` should be considered an error, precisely what radare2 does.

**Comment by typoon** [comment]
I get where you are coming from, and that exact argument should allow overwriting the `*_t` types, since they are defined in terms of the real C primitive types (int, long, char, etc.). Since there is a set of configurations that specify the size of the base primitives, the `_t` ones should then be defined based on them, but if the user has a reason to overwrite those, they should be allowed to do so. If that behavior is not desired, there should at least be a warning msg pointing the user to those configurations and radare should still parse the rest of the header instead of stoping.

**Comment by typoon** [comment]
Just to make clear what I am trying to say (was on my phone hen wrote the previous reply so it was a bit short).

The `*_t` types should probably be allowed to be rewritten exactly for the reason that the user might be loading a binary of another platform or they might have a reason to be overwriting those values.
creating any of the types that were processed up to that point.

Whenever a redefinition of radare's primitive types is encountered, perhaps display a warning showing what type is being redefined and in what file and line number that redefinition has been found.

Ideally there should be a configuration then, for example `e types.overwrite` or something, that the user can set to values such:

* all - Whenever a type that already exists is found while processing a header, the existing type in radare's database will be overwritten by the current one being processed.
* empty - Whenever a type definition that already exists and is empty (for example, an empty struct or forward declaration) is found while processing a header, the existing type in radare's database will be overwritten by the current one being processed.
* never - Whenever a type definition that already exists is found while processing a header, the new definition being processed will be ignored and a warning is presented to the user.

How does that sound?
